### PR TITLE
feat: clone datasets when creating a collection revision

### PR DESCRIPTION
**Pull request by MDunitz**
### Reviewers
**Functional:** 

**Readability:** 

---

## Changes
- add
- remove
- modify

### CODE REVIEW

File: tests/unit/backend/corpora/common/entities/datasets/test_revision.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+class TestDatasetRevision(TestDataset):
+    def test__create_dataset_revision(self):
+        dataset = self.generate_dataset_with_s3_resources(self.session, published=True)
+        rev_dataset = dataset.create_revision().to_dict()
+        dataset = dataset.to_dict()
+
+        with self.subTest("artifacts are correctly created and point to correct s3 uri"):
+            rev_artifacts = rev_dataset.pop("artifacts")
+            original_artifacts = dataset.pop("artifacts")
+            for i in range(0, len(rev_artifacts)):
```
Comment (Bento007) [inline tests/unit/backend/corpora/common/entities/datasets/test_revision.py:13]: Line 13-18 is repeated a few times below. Could it DRYed out?

File: tests/unit/backend/corpora/common/entities/datasets/test_revision.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        with self.subTest("Test processing status copied over"):
+            rev_processing_status = rev_dataset.pop("processing_status")
+            original_processing_status = dataset.pop("processing_status")
+            for key in rev_processing_status.keys():
+                if key in ("dataset_id", "id"):
+                    self.assertNotEqual(rev_processing_status[key], original_processing_status[key])
+                else:
+                    self.assertEqual(rev_processing_status[key], original_processing_status[key])
+
+        with self.subTest("revision count is correctly incremented"):
```
Comment (Bento007) [inline tests/unit/backend/corpora/common/entities/datasets/test_revision.py:38]: revision count should be equal.

**Review by Bento007 — CHANGES_REQUESTED** [review]
Don't increment revision when cloning.

**Review by Bento007 — APPROVED** [review]
LGTM

File: backend/corpora/lambdas/api/v1/collection.py [hunk 3]
```diff
@@ -58,7 +58,10 @@ def post_collection_revision(collection_uuid: str, user: str):
     except sqlalchemy.exc.IntegrityError as ex:
         db_session.rollback()
         raise ConflictException() from ex
+    for dataset in collection.datasets:
```
Comment (MDunitz) [inline backend/corpora/lambdas/api/v1/collection.py:61]:
```suggestion
```

File: backend/corpora/lambdas/api/v1/collection.py [hunk 4]
```diff
@@ -58,7 +58,10 @@ def post_collection_revision(collection_uuid: str, user: str):
...
+        Dataset(dataset).create_revision()
```
Comment (MDunitz) [inline backend/corpora/lambdas/api/v1/collection.py:62]:
```suggestion
```

File: backend/corpora/lambdas/api/v1/collection.py [hunk 5]
```diff
@@ -4,7 +4,7 @@
 from flask import make_response, jsonify, g
 
 from ....common.corpora_orm import DbCollection, CollectionVisibility
-from ....common.entities import Collection
+from ....common.entities import Collection, Dataset
```
Comment (MDunitz) [inline backend/corpora/lambdas/api/v1/collection.py:7]:
```suggestion
```

**Review by Bento007 — APPROVED** [review]

**Pull request merged**